File: apps/web/src/drizzle/schema/lessonAsset.ts
import {
  pgTable,
  text,
  uuid,
  integer,
  pgEnum,
  bigint,
  boolean,
} from "drizzle-orm/pg-core";
import { createdAt, id, updatedAt } from "../schemaHelpers";
import { relations } from "drizzle-orm";
import { LessonTable } from "./lesson";

export const assetTypes = [
  "youtube",
  "video_file",
  "pdf",
  "image",
  "audio",
] as const;
export type AssetType = (typeof assetTypes)[number];
export const assetTypeEnum = pgEnum("asset_type", assetTypes);

// where it's physically stored / how it's delivered
export const assetProviders = ["youtube", "r2", "bunny"] as const;
export type AssetProvider = (typeof assetProviders)[number];
export const assetProviderEnum = pgEnum("asset_provider", assetProviders);

// primary = the lesson's main content (what renders in the player/viewer)
// attachment = a downloadable extra alongside the primary content (slides, worksheet, source files)
export const assetRoles = ["primary", "attachment"] as const;
export type AssetRole = (typeof assetRoles)[number];
export const assetRoleEnum = pgEnum("asset_role", assetRoles);

export const LessonAssetTable = pgTable("lesson_assets", {
  id: id(),
  lessonId: uuid()
    .notNull()
    .references(() => LessonTable.id, { onDelete: "cascade" }),
  type: assetTypeEnum().notNull(),
  provider: assetProviderEnum().notNull(),
  role: assetRoleEnum().notNull().default("primary"),

  // one of these two depending on provider
  externalId: text(), // youtube video id
  storageKey: text(), // r2/bunny object key (private — used to generate signed URLs)

  fileName: text(), // original filename, for download prompts
  mimeType: text(),
  fileSizeBytes: bigint({ mode: "number" }),

  downloadable: boolean().notNull().default(false),

  durationSeconds: integer(),

  order: integer().notNull().default(0), // ordering among attachments
  createdAt,
  updatedAt,
});

export const LessonAssetRelationships = relations(
  LessonAssetTable,
  ({ one }) => ({
    lesson: one(LessonTable, {
      fields: [LessonAssetTable.lessonId],
      references: [LessonTable.id],
    }),
  }),
);

File: apps/mobile/lib/features/lessons/data/lesson_asset.dart
// lib/features/lessons/data/lesson_asset.dart
enum AssetType { youtube, videoFile, pdf, image, audio, unknown }
enum AssetRole { primary, attachment }

AssetType _parseType(String? s) {
  switch (s) {
    case 'youtube': return AssetType.youtube;
    case 'video_file': return AssetType.videoFile;
    case 'pdf': return AssetType.pdf;
    case 'image': return AssetType.image;
    case 'audio': return AssetType.audio;
    default: return AssetType.unknown;
  }
}

class LessonAsset {
  final String id;
  final AssetType type;
  final AssetRole role;
  final String? externalId; // youtube video id
  final String? fileName;
  final int? fileSizeBytes;
  final bool downloadable;
  final int? durationSeconds;

  LessonAsset({
    required this.id,
    required this.type,
    required this.role,
    this.externalId,
    this.fileName,
    this.fileSizeBytes,
    required this.downloadable,
    this.durationSeconds,
  });

  factory LessonAsset.fromJson(Map<String, dynamic> json) {
    return LessonAsset(
      id: json['id'] as String,
      type: _parseType(json['type'] as String?),
      role: (json['role'] as String?) == 'attachment' ? AssetRole.attachment : AssetRole.primary,
      externalId: json['externalId'] as String?,
      fileName: json['fileName'] as String?,
      fileSizeBytes: json['fileSizeBytes'] as int?,
      downloadable: json['downloadable'] as bool? ?? false,
      durationSeconds: json['durationSeconds'] as int?,
    );
  }
}